dbt/adapters/spark/session.py

```diff
@@ -0,0 +1,164 @@
+"""Spark session integration."""
+
+from __future__ import annotations
+
+from types import TracebackType
+from typing import Any
+
+from pyspark.sql import DataFrame, Row, SparkSession
+
+
+class Cursor:
+    """
+    Mock a pyodbc cursor.
+
+    Source
+    ------
+    https://github.com/mkleehammer/pyodbc/wiki/Cursor
+    """
+
+    def __init__(self) -> None:
+        self._df: DataFrame | None = None
+        self._rows: list[Row] | None = None
+
+    def __enter__(self) -> Cursor:
+        return self
+
+    def __exit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_val: Exception | None,
+        exc_tb: TracebackType | None,
+    ) -> bool:
+        self.close()
+        return True
+
+    @property
+    def description(
+        self,
+    ) -> list[tuple[str, str, None, None, None, None, bool]]:
+        """
+        Get the description.
+
+        Returns
+        -------
+        out : list[tuple[str, str, None, None, None, None, bool]]
+            The description.
+
+        Source
+        ------
+        https://github.com/mkleehammer/pyodbc/wiki/Cursor#description
+        """
+        if self._df is None:
+            description = list()
+        else:
+            description = [
+                (
+                    field.name,
+                    field.dataType.simpleString(),
+                    None,
+                    None,
+                    None,
+                    None,
+                    field.nullable,
+                )
+                for field in self._df.schema.fields
+            ]
+        return description
+
+    def close(self) -> None:
+        """
+        Close the connection.
+
+        Source
+        ------
+        https://github.com/mkleehammer/pyodbc/wiki/Cursor#close
+        """
+        self._df = None
+        self._rows = None
+
+    def execute(self, sql: str, *parameters: Any) -> None:
+        """
+        Execute a sql statement.
+
+        Parameters
+        ----------
+        sql : str
+            Execute a sql statement.
+        *parameters : Any
+            The parameters.
+
+        Raises
+        ------
+        NotImplementedError
+            If there are parameters given. We do not format sql statements.
+
+        Source
+        ------
+        https://github.com/mkleehammer/pyodbc/wiki/Cursor#executesql-parameters
+        """
+        if len(parameters) > 0:
+            sql = sql % parameters
+        spark_session = SparkSession.builder.getOrCreate()
+        self._df = spark_session.sql(sql)
```

Comment:
👀 really interesting how different this looks, since it's just creating a DataFrame in memory (?)

Reply:
Kinda, spark is lazy. So it is only loaded into memory at the moment `collect` in `fetchall` is called.